In [2]:
import sys
from pathlib import Path

project_root = Path.cwd().parent

sys.path.append(str(project_root))

print(project_root)

c:\Users\rishi\n100-financial-intelligence


# Sprint 1 Day 2

## Goal

Load all datasets and inspect structure.

## Datasets Loaded

- companies.xlsx
- profitandloss.xlsx
- balancesheet.xlsx
- cashflow.xlsx
- analysis.xlsx
- documents.xlsx
- prosandcons.xlsx
- financial_ratios.xlsx
- market_cap.xlsx
- peer_groups.xlsx
- sectors.xlsx
- stock_prices.xlsx

## Result

All datasets loaded successfully.

In [3]:
from src.etl.extract import load_all_datasets

datasets = load_all_datasets()

for name, df in datasets.items():
    print(name, df.shape)

companies.xlsx (92, 12)
profitandloss.xlsx (1276, 15)
balancesheet.xlsx (1312, 13)
cashflow.xlsx (1187, 7)
analysis.xlsx (20, 6)
documents.xlsx (1585, 4)
prosandcons.xlsx (16, 4)
financial_ratios.xlsx (1184, 16)
market_cap.xlsx (552, 9)
peer_groups.xlsx (56, 4)
sectors.xlsx (92, 6)
stock_prices.xlsx (5520, 9)


In [4]:
df = datasets["profitandloss.xlsx"]

df.head()

,id,company_id,year,sales,expenses,operating_profit,opm_percentage,other_income,interest,depreciation,profit_before_tax,tax_percentage,net_profit,eps,dividend_payout
0,61,ABB,Dec 2012,1653,1451,202.0,12.0,33,0,19,215,33.0,145,68.0,25.0
1,62,ABB,Mar 2014,2276,2009,267.0,12.0,49,0,22,295,33.0,198,93.0,25.0
2,63,ABB,Mar 2015,2289,1977,312.0,14.0,48,0,15,344,34.0,229,108.0,29.0
3,64,ABB,Mar 2016,2614,2250,365.0,14.0,50,3,14,398,36.0,255,120.0,29.0
4,65,ABB,Mar 2017,2903,2505,398.0,14.0,57,2,16,436,37.0,277,130.0,31.0


In [5]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1276 entries, 0 to 1275
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 1276 non-null   int64  
 1   company_id         1276 non-null   str    
 2   year               1276 non-null   str    
 3   sales              1276 non-null   int64  
 4   expenses           1276 non-null   int64  
 5   operating_profit   1263 non-null   float64
 6   opm_percentage     1261 non-null   float64
 7   other_income       1276 non-null   int64  
 8   interest           1276 non-null   int64  
 9   depreciation       1276 non-null   int64  
 10  profit_before_tax  1276 non-null   int64  
 11  tax_percentage     1181 non-null   float64
 12  net_profit         1276 non-null   int64  
 13  eps                1271 non-null   float64
 14  dividend_payout    1173 non-null   float64
dtypes: float64(5), int64(8), str(2)
memory usage: 167.9 KB


In [6]:
df.isnull().sum()

id                     0
company_id             0
year                   0
sales                  0
expenses               0
operating_profit      13
opm_percentage        15
other_income           0
interest               0
depreciation           0
profit_before_tax      0
tax_percentage        95
net_profit             0
eps                    5
dividend_payout      103
dtype: int64

## Observations

1. Dataset contains 1276 records.
2. Year column is stored as text.
3. Operating profit contains missing values.
4. OPM percentage contains missing values.
5. Tax percentage contains missing values.
6. Further validation required before database loading.

In [7]:
df[df["operating_profit"].isnull()]

,id,company_id,year,sales,expenses,operating_profit,opm_percentage,other_income,interest,depreciation,profit_before_tax,tax_percentage,net_profit,eps,dividend_payout
905,1004,PNB,Mar 2013,43071,12456,NaN,NaN,4310,27802,336,6786,27.0,5020,28.0,19.0
906,1005,PNB,Mar 2014,44958,16112,NaN,NaN,4710,28220,367,4969,29.0,3674,20.0,10.0
907,1006,PNB,Mar 2015,48710,18760,NaN,NaN,6175,31343,384,4397,24.0,3525,18.0,18.0
908,1007,PNB,Mar 2016,50804,27453,NaN,NaN,6100,34331,413,-5293,-31.0,-3510,-19.0,-2.0
909,1008,PNB,Mar 2017,48058,22533,NaN,NaN,9168,32722,430,1540,41.0,1239,6.0,0.0
910,1009,PNB,Mar 2018,48725,43345,NaN,NaN,8883,33530,581,-19849,-37.0,-12111,-44.0,0.0
911,1010,PNB,Mar 2019,52147,39645,NaN,NaN,7367,34656,584,-15370,-35.0,-9550,-21.0,0.0
912,1011,PNB,Mar 2020,54918,25867,NaN,NaN,9388,36997,614,827,56.0,485,1.0,0.0
913,1012,PNB,Mar 2021,81935,38601,NaN,NaN,12777,50805,982,4325,38.0,2695,2.0,0.0
914,1013,PNB,Mar 2022,76242,36026,NaN,NaN,12329,46823,896,4826,19.0,3908,4.0,18.0


In [8]:
df[df["eps"].isnull()]

,id,company_id,year,sales,expenses,operating_profit,opm_percentage,other_income,interest,depreciation,profit_before_tax,tax_percentage,net_profit,eps,dividend_payout
687,786,JIOFIN,Mar 2023,45,6,39.0,88.0,10,0,0,49,37.0,31,NaN,0.0
729,828,LICI,Mar 2019,571508,565587,5920.0,1.0,2615,0,336,8199,68.0,2627,NaN,2.0
730,829,LICI,Mar 2020,628043,634989,-6947.0,-1.0,20870,0,379,13544,80.0,2710,NaN,2.0
735,834,LICI,TTM,896106,870255,25851.0,3.0,16250,0,0,42102,NaN,41468,NaN,NaN
1009,1108,SIEMENS,Sep 2011,11955,10784,1171.0,10.0,299,26,154,1290,33.0,868,NaN,24.0


In [9]:
df[df["dividend_payout"].isnull()]

,id,company_id,year,sales,expenses,operating_profit,opm_percentage,other_income,interest,depreciation,profit_before_tax,tax_percentage,net_profit,eps,dividend_payout
12,73,ABB,TTM,6066,4519,1547.0,26.0,265,11,72,1728,NaN,1285,605.0,NaN
13,86,ADANIENSOL,Mar 2014,0,0,0.0,NaN,0,0,0,0,0.0,0,0.0,NaN
24,97,ADANIENSOL,TTM,22099,15507,6592.0,NaN,-844,3183,1912,653,NaN,589,6.0,NaN
37,110,ADANIENT,TTM,102311,88418,13893.0,14.0,1489,4150,3541,7691,NaN,6086,49.0,NaN
59,132,ADANIPORTS,TTM,28443,11278,17165.0,60.0,1039,2591,4053,11560,NaN,9743,45.0,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1228,1327,VEDL,TTM,144448,107376,37072.0,26.0,5266,9721,10958,21659,NaN,15841,30.0,NaN
1241,1340,WIPRO,TTM,88792,71019,17773.0,20.0,3349,1431,3077,16615,NaN,12488,12.0,NaN
1249,1348,ZOMATO,TTM,17972,17321,651.0,4.0,944,118,716,761,NaN,663,1.0,NaN
1262,1361,ZYDUSLIFE,TTM,21484,15205,6278.0,29.0,351,112,849,5668,NaN,4439,43.0,NaN


## Missing Value Investigation

Missing values detected in:

- operating_profit
- opm_percentage
- tax_percentage
- eps
- dividend_payout

Next step is to identify affected companies and years before deciding on treatment.

## Missing Operating Profit Investigation

All missing operating_profit and opm_percentage values belong to Punjab National Bank (PNB).

Observation:
Banking institutions often do not report operating profit and operating profit margin in the same manner as non-financial companies.

Decision:
Treat these missing values as "Not Applicable" rather than data quality errors.

## EPS Missing Value Investigation

Affected Companies:
- JIOFIN
- LICI
- SIEMENS

Observation:
Missing EPS values appear to be limited to a small number of records and may be due to reporting differences, listing timelines, or unavailable disclosures.

Decision:
Retain missing values until further business validation is completed.

In [10]:
df["year"].unique()

<ArrowStringArray>
[   'Dec 2012',    'Mar 2014',    'Mar 2015',    'Mar 2016',    'Mar 2017',
    'Mar 2018',    'Mar 2019',    'Mar 2020',    'Mar 2021',    'Mar 2022',
    'Mar 2023',    'Mar 2024',         'TTM',    'Mar 2013',    'Dec 2013',
    'Dec 2014',    'Dec 2015',    'Dec 2016',    'Dec 2017',    'Dec 2018',
    'Dec 2019',    'Dec 2020',    'Dec 2021', 'Mar 2023 15',    'Jun 2013',
    'Jun 2014',    'Jun 2015', 'Mar 2016 9m',    'Mar 2011',    'Mar 2012',
    'Dec 2022',    'Dec 2023',    'Sep 2011',    'Sep 2014',    'Sep 2015',
    'Sep 2016',    'Sep 2017',    'Sep 2018',    'Sep 2019',    'Sep 2020',
    'Sep 2021',    'Sep 2022',    'Sep 2023',    'Sep 2024']
Length: 44, dtype: str

In [11]:
df["year"].str.extract(r"(\d{4})")

,0
0,2012
1,2014
2,2015
3,2016
4,2017
...,...
1271,2021
1272,2022
1273,2023
1274,2024


In [12]:
from src.etl.transform import extract_year

df2 = extract_year(df)

df2[
    ["year", "financial_year"]
].head(15)

,year,financial_year
0,Dec 2012,2012
1,Mar 2014,2014
2,Mar 2015,2015
3,Mar 2016,2016
4,Mar 2017,2017
5,Mar 2018,2018
6,Mar 2019,2019
7,Mar 2020,2020
8,Mar 2021,2021
9,Mar 2022,2022


## Year Standardization

Objective:
Create a consistent financial year field for analysis.

Method:
Extracted four-digit year values from reporting labels using regex.

Observations:
- Standard annual records successfully mapped.
- TTM records remain null.
- Special reporting periods (9M, 15M) retain their original label in the year column.

Result:
New financial_year column created for future analytics and database storage.

## ETL Transformation 1 - Financial Year Extraction

Problem:
The year column contains mixed formats such as:
- Mar 2024
- Dec 2023
- Sep 2022
- TTM

Solution:
Extract four-digit years into a dedicated financial_year column.

Outcome:
The dataset can now be filtered, grouped, and analyzed by year.

In [13]:
df["company_id"].nunique()

100

In [14]:
sorted(df["company_id"].unique())

['ABB',
 'ADANIENSOL',
 'ADANIENT',
 'ADANIGREEN',
 'ADANIPORTS',
 'ADANIPOWER',
 'AMBUJACEM',
 'APOLLOHOSP',
 'ASIANPAINT',
 'ATGL',
 'AXISBANK',
 'BAJAJ-AUTO',
 'BAJAJFINSV',
 'BAJAJHLDNG',
 'BAJFINANCE',
 'BANKBARODA',
 'BEL',
 'BHARTIARTL',
 'BHEL',
 'BOSCHLTD',
 'BPCL',
 'BRITANNIA',
 'CANBK',
 'CHOLAFIN',
 'CIPLA',
 'COALINDIA',
 'DABUR',
 'DIVISLAB',
 'DLF',
 'DMART',
 'DRREDDY',
 'EICHERMOT',
 'GAIL',
 'GODREJCP',
 'GRASIM',
 'HAL',
 'HAVELLS',
 'HCLTECH',
 'HDFCBANK',
 'HDFCLIFE',
 'HEROMOTOCO',
 'HINDALCO',
 'HINDUNILVR',
 'ICICIBANK',
 'ICICIGI',
 'ICICIPRULI',
 'INDIGO',
 'INDUSINDBK',
 'INFY',
 'IOC',
 'IRCTC',
 'IRFC',
 'ITC',
 'JINDALSTEL',
 'JIOFIN',
 'JSWENERGY',
 'JSWSTEEL',
 'KOTAKBANK',
 'LICI',
 'LODHA',
 'LT',
 'LTIM',
 'M&M',
 'MARUTI',
 'MOTHERSON',
 'NAUKRI',
 'NESTLEIND',
 'NHPC',
 'NTPC',
 'ONGC',
 'PFC',
 'PIDILITIND',
 'PNB',
 'POWERGRID',
 'RECLTD',
 'RELIANCE',
 'SBILIFE',
 'SBIN',
 'SHREECEM',
 'SHRIRAMFIN',
 'SIEMENS',
 'SUNPHARMA',
 'TATACONSUM',
 'TAT

In [15]:
len(sorted(df["company_id"].unique()))

100

In [16]:
companies = datasets["companies.xlsx"]

print(
    companies["id"].nunique()
)

92


In [17]:
pnl_companies = set(
    df["company_id"].unique()
)

master_companies = set(
    companies["id"].unique()
)

extra_companies = pnl_companies - master_companies

print(extra_companies)

{'ZYDUSLIFE', 'VEDL', 'ZOMATO', 'ULTRACEMCO', 'VBL', 'UNITDSPR', 'WIPRO', 'UNIONBANK'}


## Data Quality Finding #1

Observation:

Profit & Loss dataset contains 100 unique companies.

Master company dataset contains 92 unique companies.

Extra companies identified:

- ZYDUSLIFE
- VEDL
- ZOMATO
- ULTRACEMCO
- VBL
- UNITDSPR
- WIPRO
- UNIONBANK

Interpretation:

The project universe appears intentionally restricted to 92 companies while historical financial datasets still contain records for all 100 Nifty constituents.

Action:

Validate project scope before removing records.

In [18]:
companies["id"].tolist()

['ABB',
 'ADANIENSOL',
 'ADANIENT',
 'ADANIGREEN',
 'ADANIPORTS',
 'ADANIPOWER',
 'AMBUJACEM',
 'APOLLOHOSP',
 'ASIANPAINT',
 'ATGL',
 'AXISBANK',
 'BAJAJ-AUTO',
 'BAJAJFINSV',
 'BAJAJHLDNG',
 'BAJFINANCE',
 'BANKBARODA',
 'BEL',
 'BHARTIARTL',
 'BHEL',
 'BOSCHLTD',
 'BPCL',
 'BRITANNIA',
 'CANBK',
 'CHOLAFIN',
 'CIPLA',
 'COALINDIA',
 'DABUR',
 'DIVISLAB',
 'DLF',
 'DMART',
 'DRREDDY',
 'EICHERMOT',
 'GAIL',
 'GODREJCP',
 'GRASIM',
 'HAL',
 'HAVELLS',
 'HCLTECH',
 'HDFCBANK',
 'HDFCLIFE',
 'HEROMOTOCO',
 'HINDALCO',
 'HINDUNILVR',
 'ICICIBANK',
 'ICICIGI',
 'ICICIPRULI',
 'INDIGO',
 'INDUSINDBK',
 'INFY',
 'IOC',
 'IRCTC',
 'IRFC',
 'ITC',
 'JINDALSTEL',
 'JIOFIN',
 'JSWENERGY',
 'JSWSTEEL',
 'KOTAKBANK',
 'LICI',
 'LODHA',
 'LT',
 'LTIM',
 'M&M',
 'MARUTI',
 'MOTHERSON',
 'NAUKRI',
 'NESTLEIND',
 'NHPC',
 'NTPC',
 'ONGC',
 'PFC',
 'PIDILITIND',
 'PNB',
 'POWERGRID',
 'RECLTD',
 'RELIANCE',
 'SBILIFE',
 'SBIN',
 'SHREECEM',
 'SHRIRAMFIN',
 'SIEMENS',
 'SUNPHARMA',
 'TATACONSUM',
 'TAT

## Sprint 1 Summary So Far

### Dataset Loading

Successfully loaded all 12 project datasets into pandas DataFrames.

### Dataset Validation

Verified row counts against project specification.

### Missing Value Investigation

Investigated missing values in:

* operating_profit
* opm_percentage
* tax_percentage
* eps
* dividend_payout

Conclusion:
Most missing values appear to be business-context driven rather than data corruption.

### Company Coverage Investigation

Found 100 unique companies in Profit & Loss data versus 92 companies in the master dataset.

Extra companies:

* WIPRO
* ZOMATO
* VEDL
* ULTRACEMCO
* VBL
* UNITDSPR
* UNIONBANK
* ZYDUSLIFE

Conclusion:
Difference appears to be caused by project scope filtering rather than ETL issues.

### ETL Transformation

Created financial_year extraction logic from reporting labels.

Examples:

* Mar 2024 → 2024
* Dec 2023 → 2023
* TTM → Null

### Status

Sprint 1 ETL exploration and validation phase completed.
